### PR TITLE
Change Interface of  NDArray & TBlob for DLPack Compatible

### PR DESCRIPTION
**Pull request by ZihengJiang**
@tqchen @piiswrong @mli

### CODE REVIEW

File: include/mxnet/ndarray.h [hunk 1]
```diff
@@ -405,6 +402,16 @@ class NDArray {
     }
   };
 
+  void SetTBlob() {
```
Comment (piiswrong) [inline include/mxnet/ndarray.h]: Do you still need this if you are setting it everytime in data()?

File: include/mxnet/tensor_blob.h [hunk 2]
```diff
@@ -213,6 +214,14 @@ class TBlob {
     return static_cast<DType*>(dptr_);
   }
   /*!
+   * \brief return the corresponding DLTensor
+   * \return the address of internal DLTensor
+   */
+  inline const DLTensor& GetDLTensor() {
```
Comment (piiswrong) [inline include/mxnet/tensor_blob.h]: GetDLTensor -> dltensor

File: src/ndarray/ndarray.cc [hunk 3]
```diff
@@ -614,8 +614,7 @@ NDArray &NDArray::operator/=(const real_t &src) {
 }
 
 void NDArray::Save(dmlc::Stream *strm) const {
-  // save shape
-  shape_.Save(strm);
```
Comment (piiswrong) [inline src/ndarray/ndarray.cc]: @tqchen Does ndarray dupm files have a version system? Can we store new arrays as int64 while supporting legacy read from int32?

Comment (mli) [inline src/ndarray/ndarray.cc]: i don't think so. but we should add a version number

Comment (ZihengJiang) [inline src/ndarray/ndarray.cc]: Let us do it in another PR. For now, we just make sure it is backward compatible.

Comment (tqchen) [inline src/ndarray/ndarray.cc]: see https://github.com/dmlc/mxnet/blob/master/src/ndarray/ndarray.cc#L688

Comment (tqchen) [inline src/ndarray/ndarray.cc]: There is a reserved field on the dump file format that can be used for version checking

Comment (tqchen) [inline src/ndarray/ndarray.cc]: @ZihengJiang maybe consider to do a backward compatible version now by utilizing the hint?

Comment (piiswrong) [inline src/ndarray/ndarray.cc]: You can bump the version to 0.9.7 and store that in the reserved field

Comment (ZihengJiang) [inline src/ndarray/ndarray.cc]: ok, no problem

File: include/mxnet/tensor_blob.h [hunk 4]
```diff
@@ -281,6 +291,34 @@ class TBlob {
...
+        {3, {1,  8, 1}},  // UInt8
+        {4, {0, 32, 1}},  // Int32
+        {5, {0,  8, 1}}   // Int8
+      };
+    return MSHADOW_DTYPE_TO_DLPACK_DTYPE[type_flag];
+  }
+
+  inline void SetDLTensor() {
+    dltensor_.data = dptr_;
+    dltensor_.ctx = DLContext{-1, static_cast<DLDeviceType>(dev_mask_)};
```
Comment (tqchen) [inline include/mxnet/tensor_blob.h]: we need to fix context to be the right one when we really want to use DLTensor, otherwise it won't be usable

Comment (ZihengJiang) [inline include/mxnet/tensor_blob.h]: Then how to handle the legacy constructors without device id?

Comment (tqchen) [inline include/mxnet/tensor_blob.h]: @piiswrong any ideas? do you think if we can construct TBlob with dev_id by default?

Comment (piiswrong) [inline include/mxnet/tensor_blob.h]: we can have a new argument defaults to -1
then modify all the places it was used in mxnet

Comment (tqchen) [inline include/mxnet/tensor_blob.h]: to be clear, device_id equals -1 won't be able to be used in dlpack compatible functions

File: include/mxnet/tensor_blob.h [hunk 5]
```diff
@@ -67,36 +63,37 @@ class TBlob {
...
    */
-  TBlob(void *dptr,
-        const TShape &shape,
-        int dev_mask,
-        int type_flag)
-      : dptr_(dptr), shape_(shape),
-        stride_(shape[shape.ndim() - 1]),
-        dev_mask_(dev_mask),
-        type_flag_(type_flag) {
+  TBlob(void *dptr, const TShape &shape, int dev_mask, int dev_id, int type_flag)
```
Comment (piiswrong) [inline include/mxnet/tensor_blob.h]: put dev_id at last and give it a default value -1

File: include/mxnet/ndarray.h [hunk 6]
```diff
@@ -121,7 +121,7 @@ class NDArray {
     raw_shape[0] = length;
     MSHADOW_TYPE_SWITCH(dtype_, DType, {
       res = TBlob(static_cast<DType*>(ptr_->shandle.dptr)
-        + offset_ + offset, raw_shape, ptr_->shandle.ctx.dev_mask());
+        + offset_ + offset, raw_shape, ptr_->shandle.ctx);
```
Comment (tqchen) [inline include/mxnet/ndarray.h]: why two offset here? maybe consider combine into one?

File: .gitmodules [hunk 7]
```diff
@@ -9,7 +9,10 @@
 	url = https://github.com/dmlc/ps-lite
 [submodule "nnvm"]
 	path = nnvm
-	url = https://github.com/dmlc/nnvm
+	url = https://github.com/ZihengJiang/nnvm
```
Comment (mli) [inline .gitmodules]: why use `ZihengJiang/nnvm` instead of `dmlc/nnvm`?

Comment (ZihengJiang) [inline .gitmodules]: As you can see here in the log of ci(http://ec2-52-25-96-65.us-west-2.compute.amazonaws.com/blue/organizations/jenkins/mxnet/detail/PR-6345/12/pipeline/220#step-293-log-131). The size of input_shape is right in the mxnet side but wrong in the nnvm side. So I'm trying to log it out in my nnvm branch for debuging. (There maybe something wrong in CI with updated submodule). But it looks like this method doesn't work due to CI cannot change to my branch.

File: include/mxnet/ndarray.h (hunk 6 above)
Comment (ZihengJiang) [inline include/mxnet/ndarray.h]: the original tblob::chunk has one offset, also raw_data can accept an offset as an argument

File: src/ndarray/ndarray.cc [hunk 8]
```diff
@@ -639,25 +643,35 @@ void NDArray::Save(dmlc::Stream *strm) const {
 }
 
 bool NDArray::Load(dmlc::Stream *strm) {
```
Comment (piiswrong) [inline src/ndarray/ndarray.cc]: change the interface directly

**Comment by piiswrong** [comment]
please add tests for legacy ndarray loading

File: src/ndarray/ndarray.cc (hunk 8 above)
Comment (ZihengJiang) [inline src/ndarray/ndarray.cc]: then it requires to change here: https://github.com/ZihengJiang/mxnet/blob/tvm/include/mxnet/ndarray.h#L337
ok?

File: include/mxnet/tensor_blob.h [hunk 9]
```diff
@@ -54,49 +49,61 @@ class TBlob {
...
    * \param dptr the pointer to the memory
    * \param shape the shape of the data
    * \param dev_mask the device mask, can be cpu::kDevMask or gpu::kDevMask
+   * \param dev_id the device id
    */
   template<typename DType>
-  TBlob(DType *dptr,
-        const TShape &shape,
-        int dev_mask)
+  TBlob(DType *dptr, const TShape &shape, int dev_mask, int dev_id)
```
Comment (piiswrong) [inline include/mxnet/tensor_blob.h]: TBlob constructor interface need to be backward compatible
